internal/datastore/crdb/pool/pool.go

```diff
@@ -0,0 +1,393 @@
+func (p *RetryPool) Close() {
+	p.pool.Close()
+}
+
+// Stat returns the underlying pgxpool.Pool stats
+func (p *RetryPool) Stat() *pgxpool.Stat {
+	return p.pool.Stat()
+}
+
+// Node returns the id for a connection
+func (p *RetryPool) Node(conn *pgx.Conn) uint32 {
+	p.Lock()
+	defer p.Unlock()
+	id, ok := p.nodeForConn[conn]
+	if !ok {
+		return 0
+	}
+	return id
+}
+
+// Range applies a function to every entry in the connection list
+func (p *RetryPool) Range(f func(conn *pgx.Conn, nodeID uint32)) {
+	p.Lock()
+	defer p.Unlock()
+	for k, v := range p.nodeForConn {
+		f(k, v)
+	}
+}
+
+// withRetries acquires a connection and attempts the request multiple times
+func (p *RetryPool) withRetries(ctx context.Context, fn func(conn *pgxpool.Conn) error) error {
+	conn, err := p.pool.Acquire(ctx)
+	if err != nil {
+		conn.Release()
+		return fmt.Errorf("error acquiring connection from pool: %w", err)
+	}
+	defer func() {
+		conn.Release()
+	}()
+
+	var retries uint32
+	defer func() {
+		resetHistogram.Observe(float64(retries))
+	}()
+
+	maxRetries := p.maxRetries
+	if disabled, ok := ctx.Value(CtxDisableRetries).(bool); ok && disabled {
+		maxRetries = 0
+	}
+
+	for retries = uint32(0); retries <= maxRetries; retries++ {
+		err = wrapRetryableError(ctx, fn(conn))
+		if err == nil {
+			conn.Release()
+			if retries > 0 {
+				log.Ctx(ctx).Info().Uint32("retries", retries).Msg("resettable database error succeeded after retry")
+			}
+			return nil
+		}
+
+		var (
+			resettable *ResettableError
+			retryable  *RetryableError
+		)
+		if errors.As(err, &resettable) {
+			log.Ctx(ctx).Info().Err(err).Uint32("retries", retries).Msg("resettable error")
+
+			id := p.safeGC(ctx, conn)
+			conn.Release()
+
+			// After a resettable error, mark the node as unhealthy
+			// TODO: configurable error count / circuit-breaker
+			if id > 0 {
+				p.healthTracker.SetNodeHealth(id, false)
+			}
+
+			sleepOnErr(ctx, err, retries)
+
+			conn, err = p.acquireFromDifferentNode(ctx, id)
+			if err != nil {
+				return fmt.Errorf("error acquiring connection from pool after retry %d: %w", retries, err)
+			}
+			continue
+		}
+		if errors.As(err, &retryable) {
+			log.Ctx(ctx).Info().Err(err).Uint32("retries", retries).Msg("retryable error")
+			sleepOnErr(ctx, err, retries)
+			continue
+		}
+		conn.Release()
+		// error is not resettable or retryable
+		log.Ctx(ctx).Warn().Err(err).Uint32("retries", retries).Msg("error is not resettable or retryable")
+		return err
+	}
+	return &MaxRetryError{MaxRetries: p.maxRetries, LastErr: err}
+}
+
+// safeGC recovers if the underlying puddle's connection is nil
+// It returns a nodeID of zero if so, since we lost access to the connection object
+// to compute the id from.
+func (p *RetryPool) safeGC(ctx context.Context, conn *pgxpool.Conn) uint32 {
+	var id uint32
+	defer func() {
+		if r := recover(); r != nil {
+			log.Ctx(ctx).Warn().Msg("recovered from a missing pgx.Conn on a pgxpool.Conn")
+		}
+	}()
```

Comment:
yes - I suspect this is actually a pgx bug where the (pgx) health checker kills a connection even though I have checked out from the pool, but I haven't repro'd it in isolation. It only seems to happen when the connection is actively killed by a draining cockroach node.

Reply:
Got it. I thought of actually disabling pgx conn pool, since you replicated some of its behaviour here, but there are bits that are not exposed (like the check for minConnections)

Reply:
I spent some time to track this down and came up with: https://github.com/jackc/pgx/issues/1607 (it wasn't what I originally guessed - pgxpool health checks are fine).

I switched things around to guard against this one case, and ran it against a cluster that I kept rolling, and never saw an issue.

Not 100% sure why that behavior is there in pgx, so just worked around it for now and we can potentially remove it in the future based on the discussion over there. But for now no more `recover` 👍

Reply:
Tracked down the issue - there was a stray `release` which was the original cause of the panic. The code I wrote to work around it _missed_ a release, which is why pgxpool still had a high number of connections even though cockroach did not.

Fixed this all now and have tested across multiple upgrades in a real cluster.